### PR TITLE
THREESCALE 9308 HeadCrab Vulnerability (Redis)

### PR DESCRIPTION
**Pull request by valerymo**
Jira:  https://issues.redhat.com/browse/THREESCALE-9308

**This PR is recreated from https://github.com/3scale/3scale-operator/pull/954, that was closed because of significant changes in master**  

- Current PR is intended to prevent redis replication and slaves creation to prevent HeadCrab Vulnerability.
- for Internal Redis Only.
- If Redis is Internal than following records are added to configuration to disable replicaoff and slaveoff redis cli commands: 
      rename-command REPLICAOF ""
      rename-command SLAVEOF ""
- If both or one of these `rename-command` records removed or not found in `redis-config` config map - controller will restore them.
- PR does not prevent changes to other settings in the reids configuration (not REPLICAOF or SLAVEOF)

## Tests Preparation - files
- s3-creds-secret.yaml

```
kind: Secret
apiVersion: v1
metadata: 
  name: s3-credentials
  namespace: 3scale-test
data: 
  AWS_ACCESS_KEY_ID: QUtJQVY2SVcccc
  AWS_SECRET_ACCESS_KEY: aU5VbWdZY3hjSDFccc
  AWS_BUCKET: dm1vY2NzZjZxOGxyZWRoYXRyaGccccc
  AWS_REGION: ZXUtd2Vcccc
type: Opaque
```

- apimanagerCR.yaml

```
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: example-apimanager
spec:
  system: 
    fileStorage: 
      simpleStorageService: 
        configurationSecretRef: 
          name: s3-credentials
  wildcardDomain: apps.xxxxxx.axxxx.xxx.xxxxx.org

```

## Validation

-  get this PR

cd .../3scale-operator and run 3scale installation:

```
make install
oc new-project 3scale-test
make download
make run
```

- in another xterm:

```
oc apply -f s3-creds-secret.yaml
oc apply -f apimanagerCR.yaml
```

- Check that `redis-config` ConfigMap has required changes:
```
$ oc describe cm redis-config |grep rename-command
rename-command REPLICAOF ""
rename-command SLAVEOF ""
```

- Check that replicaof and slaveof commands are disabled.
Expected behavior as in example below

```
kubectl exec -ti $(kubectl get pods -l deployment=backend-redis -o name) -- /bin/bash
bash-4.2$ redis-cli 
127.0.0.1:6379> replicaof 10.10.10.10 6379
(error) ERR unknown command `replicaof`, with args beginning with: `10.10.10.10`, `6379`,
127.0.0.1:6379> replicaof
(error) ERR unknown command `replicaof`, with args beginning with: 
127.0.0.1:6379> slaveof
(error) ERR unknown command `slaveof`, with args beginning with: 
127.0.0.1:6379> 
```

Notes: Internally, within redis system, the `slaveof` and `replicaof` commands will be disabled. 
So the only way to enable it would be changing the configuration 
See next item  - that operator is controllilng `redis-config` config map and  will restore records that prevening HeadCrab Vulnerability.

- Check that `redis-config` config map changes are reverted
    - Edit config map - delete `rename-command REPLICAOF ""` or/and `rename-command SLAVEOF ""`
    ```
    $ oc edit cm redis-config
    ......
    ......
    // delete these lines
      rename-command REPLICAOF ""
      rename-command SLAVEOF ""
    ```
    ```
    $ oc edit cm redis-config
    configmap/redis-config edited
    ```

    - Check configMap and see that these lines are exists (reconsiler reverted configmap) 
    ```
    rename-command REPLICAOF ""
    rename-command SLAVEOF ""
    .....
    Events:  <none>
    $ oc describe cm redis-config
    ```
### Check Upgrade
**Following validation was done**
- Create Images/Indexes for master and for branch of PR

-  Install master ; check that redis-config CM does Not contain rename-command for REPLICAOF and SLAVEOF 
```
$ oc describe cm redis-config |grep rename-com
```
- Do upgrade  and check that Redis configuration was updated, new records added: 
```
$ oc describe cm redis-config |grep rename-com
rename-command REPLICAOF ""
rename-command SLAVEOF ""
```

### Regression testing
**PR does not prevent changes to other settings in the reids configuration (not REPLICAOF or SLAVEOF)**
Make sure it works as expected:  
- edit the `redis-config` configuration map, edit one or more entries/definitions.
- describe the `redis-config` configuration map and make sure the changes are in place.

**Wrong formatting of redis configuration**
In Redis, if the configuration file (redis.conf) contains a setting with an incorrect name or syntax, Redis will typically treat it as an error. Redis expects the configuration file to be correctly formatted with valid configuration directives.    
If there's a typo or an unrecognized setting name in the configuration file, Redis will log an error when starting up. - -- try change one or `rename-command` or other settings name, or add something like`test` in `redis-config` config map.  Expected behavior:

```
$ oc edit cm redis-config
....
    dir /var/lib/redis/data
    xxxrename-command SLAVEOF ""
    xxxrename-command REPLICAOF ""
kind: ConfigMap
metadata:
.......
$ oc edit cm redis-config
configmap/redis-config edited

$ oc get pod |grep redis
backend-redis-6fcbd5964f-tnp7n           0/1     CrashLoopBackOff   1 (5s ago)    6s
system-redis-77bbb6f955-d9nxs            0/1     CrashLoopBackOff   1 (5s ago)    6s

$ oc logs backend-redis-6fcbd5964f-tnp7n
*** FATAL CONFIG FILE ERROR (Redis 6.2.7) ***
Reading the configuration file, at line 47
>>> 'xxxrename-command SLAVEOF ""'
Bad directive or wrong number of arguments
$ 
```
- restore changed settings:
```
    rename-command SLAVEOF ""
    rename-command REPLICAOF ""
kind: ConfigMap
metadata:
  creationTimestamp: "2024-04-08T16:07:40Z"
$ oc edit cm redis-config
configmap/redis-config edited

$ oc get pod |grep redis
backend-redis-59bd88f884-4tvrk           1/1     Running            0             39s
system-redis-549598d55b-nk9lr            1/1     Running            0             38s

```

### CODE REVIEW

**Comment by valerymo** [comment]
- **Local Fresh install TEST**

```
$ oc get pod  |grep redis
backend-redis-76c9c446cc-6hrsj           1/1     Running     0               4m35s
system-redis-68857df9c7-zgfr2            1/1     Running     0               4m33s

$ oc rsh backend-redis-76c9c446cc-6hrsj
sh-5.1$ redis-cli
127.0.0.1:6379> replicaof
(error) ERR unknown command `replicaof`, with args beginning with: 
127.0.0.1:6379> replicaof 10.10.10.10 6379
(error) ERR unknown command `replicaof`, with args beginning with: `10.10.10.10`, `6379`, 
127.0.0.1:6379> slaveof
(error) ERR unknown command `slaveof`, with args beginning with: 
127.0.0.1:6379> 
......

$ oc rsh system-redis-68857df9c7-zgfr2
sh-5.1$ redis-cli
127.0.0.1:6379> replicaof
(error) ERR unknown command `replicaof`, with args beginning with: 
127.0.0.1:6379> slaveof
(error) ERR unknown command `slaveof`, with args beginning with: 
127.0.0.1:6379>
.....

```

- regarding comment about possible redeployment of redis pod in Local Fresh install


Looks like it happens for redis-system:
```

$ oc rollout history deploy backend-redis
deployment.apps/backend-redis 
REVISION  CHANGE-CAUSE
1         <none>

$ oc rollout history deploy system-redis 
deployment.apps/system-redis 
REVISION  CHANGE-CAUSE
1         <none>
2         <none>

```
Thank you for solution @eguzki, https://github.com/valerymo/3scale-operator/pull/5/commits, I will try add it to this PR and test it in a separate/next commit

**Comment by valerymo** [comment]
Applied and tested: https://github.com/valerymo/3scale-operator/pull/5/files - 
to avoid redeployment of redis pod in **fresh installation.**
(Changed applied manually; not git merge)

```
$ oc rollout history deploy backend-redis
deployment.apps/backend-redis 
REVISION  CHANGE-CAUSE
1         <none>

$ oc rollout history deploy system-redis
deployment.apps/system-redis 
REVISION  CHANGE-CAUSE
1         <none>

$ oc get pod |grep redis
backend-redis-6c69fd9bf9-925xg           1/1     Running     0               4m7s
system-redis-745b655b47-flk7r            1/1     Running     0               4m7s

$ oc rsh backend-redis-6c69fd9bf9-925xg
sh-5.1$ redis-cli
127.0.0.1:6379> slaveof
(error) ERR unknown command `slaveof`, with args beginning with: 
127.0.0.1:6379> replicaof 10.10.10.10 6379
(error) ERR unknown command `replicaof`, with args beginning with: `10.10.10.10`, `6379`, 
127.0.0.1:6379> replicaof
(error) ERR unknown command `replicaof`, with args beginning with: 
127.0.0.1:6379> slaveof
(error) ERR unknown command `slaveof`, with args beginning with: 
127.0.0.1:6379> 
```

**Comment by valerymo** [comment]
TODO - fix unit-tests

File: pkg/3scale/amp/component/redis.go [hunk 1]
```diff
@@ -568,3 +490,118 @@ func (redis *Redis) buildEnv() []v1.EnvVar {
...
+	}
+}
+
+func (r *RedisConfigMap) buildConfigMapData() map[string]string {
+	return map[string]string{
+		"redis.conf": r.getRedisConfData(),
+	}
+}
+
+func (r *RedisConfigMap) getRedisConfData() string { // TODO read this from a real file
```
Comment (MStokluska) [inline pkg/3scale/amp/component/redis.go:532]: @valerymo are we happy with the fact that now, not only we disable the "replica/slave of" commands but also make it impossible to add any new configuration that's different to the below?
The configuration can't be extended or edited - is this a problem in your opinion?

Comment (MStokluska) [reply pkg/3scale/amp/component/redis.go:532]: I get that but it seems like breaking current flow (that's been there for many years) - it pretty much means we are going to from now, revert everything (if there are changes present) to what we want. Let's discuss this further.

Comment (valerymo) [reply pkg/3scale/amp/component/redis.go:532]: Thank you for comment Michal. Yes, it's security restrictions,To prevent config map manipulation, and attempts to "play" with "small" changes in `rename-command` records in CM or in redis.conf itself. Sure, we can discuss at meeting. Thanks again

Comment (valerymo) [reply pkg/3scale/amp/component/redis.go:532]: It's for Internal Redis only. It's managed by Operator ..  Yes, will discuss it. Thank you

Comment (valerymo) [reply pkg/3scale/amp/component/redis.go:532]: discussed and updated, Thanks to @briangallagher  and @MStokluska

File: pkg/reconcilers/configmap.go [hunk 2]
```diff
@@ -18,3 +19,29 @@ func ConfigMapReconcileField(desired, existing *v1.ConfigMap, fieldName string)
...
+		replicaOfString := "rename-command REPLICAOF \"\""
+		if !strings.Contains(existingString, replicaOfString) {
+			newString := existingString + "\n" + replicaOfString
+			existing.Data[fieldName] = newString
+			updated = true
+		}
+
+		slaveOfString := "rename-command SLAVEOF \"\""
+		if !strings.Contains(existingString, slaveOfString) {
+			newString := existingString + "\n" + slaveOfString
```
Comment (MStokluska) [inline pkg/reconcilers/configmap.go:41]: I think this needs to be:
existingString = existingString + "\n" + <RENAME_COMMAND> in both, REPLICAOF and SLAVEOF. 
Otherwise we are going to need 2 reconciles for both to be applied if both are missing.

File: bundle/manifests/3scale-operator.clusterserviceversion.yaml [hunk 3]
```diff
@@ -515,9 +515,9 @@ spec:
                 - name: RELATED_IMAGE_SYSTEM_MEMCACHED
                   value: memcached:1.5
                 - name: RELATED_IMAGE_BACKEND_REDIS
-                  value: quay.io/centos7/redis-6-centos7:latest
+                  value: quay.io/centos7/redis-6-centos7:centos7
```
Comment (MStokluska) [inline bundle/manifests/3scale-operator.clusterserviceversion.yaml:518]: You also need to change it here: https://github.com/3scale/3scale-operator/blob/master/config/manager/manager.yaml#L64 otherwise the tests will fail on pr

Comment (MStokluska) [reply bundle/manifests/3scale-operator.clusterserviceversion.yaml:518]: also, this redis image is wrong, we should instead use: quay.io/fedora/redis-6

File: bundle/manifests/3scale-operator.clusterserviceversion.yaml [hunk 4]
```diff
@@ -515,9 +515,9 @@ spec:
...
                 - name: RELATED_IMAGE_SYSTEM_REDIS
-                  value: quay.io/centos7/redis-6-centos7:latest
+                  value: quay.io/centos7/redis-6-centos7:centos7
                 - name: RELATED_IMAGE_SYSTEM_MYSQL
                   value: centos/mysql-80-centos7
```
Comment (MStokluska) [inline bundle/manifests/3scale-operator.clusterserviceversion.yaml:522]: would you also mind updating mysql and both PostgreSQL images?

File: bundle/manifests/3scale-operator.clusterserviceversion.yaml [hunk 5]
```diff
@@ -515,15 +515,15 @@ spec:
                 - name: RELATED_IMAGE_SYSTEM_MEMCACHED
                   value: memcached:1.5
                 - name: RELATED_IMAGE_BACKEND_REDIS
-                  value: quay.io/centos7/redis-6-centos7:latest
```
Comment (eguzki) [inline bundle/manifests/3scale-operator.clusterserviceversion.yaml:518]: make these changes in another PR, please. They are not related to this PR issue

File: pkg/3scale/amp/component/redis.go [hunk 6]
```diff
@@ -568,3 +490,118 @@ func (redis *Redis) buildEnv() []v1.EnvVar {
...
+
+func (r *RedisConfigMap) buildConfigMapObjectMeta() metav1.ObjectMeta {
+	return metav1.ObjectMeta{
+		Name:      redisConfigVolumeName,
+		Labels:    r.Options.Labels,
+		Namespace: r.Options.Namespace,
+	}
+}
+
+func (redis *Redis) buildConfigMapTypeMeta() metav1.TypeMeta {
```
Comment (eguzki) [inline pkg/3scale/amp/component/redis.go:519]: this is my own leftoever, can you please remove it? thnx

File: pkg/reconcilers/configmap.go [hunk 7]
```diff
@@ -18,3 +19,29 @@ func ConfigMapReconcileField(desired, existing *v1.ConfigMap, fieldName string)
...
+
+	if existingVal, ok := existing.Data[fieldName]; !ok {
+		existing.Data[fieldName] = desired.Data[fieldName]
+		updated = true
+	} else {
+		existingString := existingVal
+
+		replicaOfString := "rename-command REPLICAOF \"\""
+		if !strings.Contains(existingString, replicaOfString) {
+			newString := existingString + "\n" + replicaOfString
```
Comment (eguzki) [inline pkg/reconcilers/configmap.go:34]: I will leave this up to you. I would not do this. I would not "modify" a file concatenating strings. It's error prone. If the `rename-command REPLICAOF \"\"` is not there, the file is invalid and hence, the operator updates it entirely.

Comment (briangallagher) [reply pkg/reconcilers/configmap.go:34]: the idea was to append to the existing config if not present rather than fully replace it. This allows for a custom configuration to continue. It's not clear from the documentation if edits to the configuration are allowed / supported but we don't prevent a user from doing so, so we can assume at least some customers have done so.

Comment (eguzki) [reply pkg/reconcilers/configmap.go:34]: LGTM then

**Review by eguzki — APPROVED** [review]

File: pkg/reconcilers/configmap.go [hunk 8]
```diff
@@ -18,3 +19,29 @@ func ConfigMapReconcileField(desired, existing *v1.ConfigMap, fieldName string)
...
 }
+
+func RedisConfigMapReconcileField(desired, existing *v1.ConfigMap, fieldName string) bool {
+	updated := false
+
+	if existingVal, ok := existing.Data[fieldName]; !ok {
+		existing.Data[fieldName] = desired.Data[fieldName]
+		updated = true
+	} else {
+		existingString := existingVal
```
Comment (eguzki) [inline pkg/reconcilers/configmap.go:30]: out of interest, what this doing?
